working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+    case Severity.warning(prefix: p):
+      print('$p: $message');
+  }
+}
 ```
 
-It's useful in places where you need a subpattern in order to destructure later
-positional values:
+The `extractorName` must refer to a named type or an enum value. (If you want to
+use an extractor with a function type, you can use a typedef.)
 
-```
-var list = [1, 2, 3];
-var [_, two, _] = list;
-```
+It is a compile-time error if:
 
-#### Variable binder
+*   `extractorName` does not refer to a type or enum value.
 
-A variable binding pattern matches the value and binds it to a new variable.
-These often occur as subpatterns of a destructuring pattern in order to capture
-a destructured value.
+*   A type argument list is present and does not match the arity of the type of
+    `extractorName` or `extractorName` is an enum value.
 
-```
-variableBinder ::= typeWithBinder? identifier
+*   A `patternField` is of the form `pattern`. Positional fields aren't allowed.
+
+As with record patterns, the getter name can be omitted in which case it is
+inferred from the variable or cast pattern in the field subpattern, which may be
+wrapped in null-check or null-assert patterns. The previous example could be
+written like:
+
+```dart
+log(Severity severity, String message) {
+  switch (severity) {
+    case Severity.error(:prefix):
+      print('!! $prefix !! $message'.toUppercase());
+    case Severity.warning(:prefix):
+      print('$prefix: $message');
+  }
+}
 ```
 
-#### Cast binder
+### Type pattern
 
-A cast pattern explicitly casts the matched value to the expected type.
+Type patterns allow extracting the runtime type arguments from objects. Extend
+the `type` grammar like so:
 
 ```
-castBinder ::= identifier "as" type
+type ::= functionType '?'?
+  | typeNotFunction
+  | typePattern // New.
+
+typePattern ::= 'final' identifier
 ```
 
-This is not a type *test* that causes a match failure if the value isn't of the
-tested type. This pattern can be used in irrefutable contexts to forcibly assert
-the expected type of some destructured value. This isn't useful as the outermost
-pattern in a declaration since you can always move the `as` to the initializer
-expression:
+Type patterns can only be used in certain contexts. It would be pretty confusing
+to allow:
 
 ```dart
-num n = 1;
-var i as int = n; // Instead of this...
-var i = n as int; // ...do this.
+print(<final E>[]);
 ```
 
-But when destructuring, there is no place in the initializer to insert the cast.
-This pattern lets you insert the cast as values are being pulled out by the
-pattern:
+It is a compile-time error if a `typePattern` appears anywhere except inside
+the `typeArguments` in:
+
+*   A `listPattern`.
+*   A `mapPattern`.
+*   A `extractorPattern`.
+*   The type annotation in a `localVariableDeclaration`.
+
+**TODO: Specify this restriction more precisely.**
+
+**TODO: Should we expand this to allow type patterns in for-loop variables,
+fields, or top-level variables?**
```

Comment:
Local declarations in general should be allowed to be patterns. That includes loop variables of a `for`/`in` and *parameters declarations*.
So, definitely allow 
```dart
for (List<final T> list in listOfLists as List<List<Object?>>) ...
```
and 
```dart
Set<S> toSet<S>(List<final T extends S> list) => <T>{...list};
```

I'd not allow patterns for "non-scoped" declarations (top-level, static, instance variables), because code accessing those are not necessarily dominated by the variable initialization. Pattern matching and binding is a scope based behavior, it only makes sense when there is a control flow that it can follow.